pkg/core/object/metadata.go

```diff
@@ -0,0 +1,115 @@
+package object
+
+import (
+	"bytes"
+	"fmt"
+	"math/big"
+	"slices"
+	"strings"
+
+	"github.com/nspcc-dev/neofs-sdk-go/client"
+)
+
+// SearchResultMerger merges SearchV2 results from several sources. Must not be
+// used concurrently.
+type SearchResultMerger struct {
+	limit, n    uint16
+	more        bool
+	buf         []client.SearchResultItem
+	iBuf, iNext *big.Int
+}
+
+// NewSearchResultMerger constructs SearchResultMerger merging up to limit
+// results.
+func NewSearchResultMerger(limit uint16) *SearchResultMerger {
+	return &SearchResultMerger{limit: limit}
+}
+
+// Add accumulates the next result.
+func (x *SearchResultMerger) Add(withCursor bool, next []client.SearchResultItem) {
+	if len(next) > int(x.limit) {
+		panic(fmt.Sprintf("limit of items exceeded %d > %d", len(next), x.limit))
+	}
+	if withCursor && len(next) < int(x.limit) {
+		panic(fmt.Sprintf("cursor is set with fewer items %d < %d", len(next), x.limit))
+	}
+	if len(next) == 0 {
+		return
+	}
+	if x.buf == nil {
+		x.buf = slices.Grow(next, int(x.limit))[:x.limit]
```

Comment:
Not very efficient, you may have no other elements. Many searches end up with 1-3 elements and allocating a 1000 for them doesn't seem to be good.